___
<img style="float: right; margin: 15px 15px 15px 15px;" src="https://cdn.brandfetch.io/idjblR01Uf/w/400/h/400/theme/dark/icon.jpeg?c=1dxbfHSJFAPEGdCLU4o5B" width="180px" height="180px" />

# <font color= #5aafc1> **PROYECTO FINAL** </font>
#### <font color= #2E9AFE> `Red Aeroportuaria Internacional`</font>
- <Strong> Sofía Maldonado, Viviana Toledo, César Adrián Santos </Strong>
- <Strong> Fecha </Strong>: 17/07/2025

___

In [38]:
# Actualizar 
import importlib
import queries 
importlib.reload(queries)
from queries import DB_Queries

# Establecer conexión
db = DB_Queries('neo5j')

Successfully Connected to Neo4j in neo5j database!
Successfully Connected to GraphDataScience in neo5j database!


# <font color= #5aafc1> **Desarrollo** </font>

Añadir atributos de Continente a cada nodo:

In [2]:
query1 = """MATCH (a:Airport) 
WHERE a.country = 'Burma'
SET a.country = 'Myanmar'
"""

query2 = """MATCH (a:Airport) 
WHERE a.country = 'Congo (Kinshasa)'
SET a.country = 'Democratic Republic of the Congo'
"""

query3 = """MATCH (a:Airport) 
WHERE a.country = 'Congo (Brazzaville)'
SET a.country = 'Republic of the Congo'
"""

query4 = f"""LOAD CSV WITH HEADERS FROM 'https://raw.githubusercontent.com/ch0fas/MG_PROYECTO2/main/data/countries.csv' AS row
MATCH (a:Airport)
WHERE a.country = row.Country
SET a.continent = row.Continent
"""

db.custom_query(query1)
db.custom_query(query2)
db.custom_query(query3)
db.custom_query(query4)

Empty DataFrame
Columns: []
Index: []

Crear subgrafos:

In [39]:
db.create_subgraph('Airport', 'TO', 'NATURAL', 'distance')
db.create_subgraph('Airport', 'TO', 'REVERSE', 'distance')
db.create_subgraph('Airport', 'TO', 'UNDIRECTED', 'distance')

Successfully created Directed Subgraph with Airport nodes and TO relationships!
Successfully created Reversed Subgraph with Airport nodes and TO relationships!
Successfully created Undirected Subgraph with Airport nodes and TO relationships!


Centrality Algorithms:

In [4]:
db.degrees('Airport')
db.page_rank('Airport')
db.betwenness('Airport')
db.closeness('Airport')

Degree Attribute added Successfully!
In_Degree Attribute added Successfully!
Out_Degree Attribute added Successfully!
Page Rank Attribute added Successfully!
Betweenness Attribute added Successfully!
Closeness Attribute added Successfully!


#### <font color= #5aafc1> **4.2 Exploratory Data Analysis** </font>

Graph Schema:

Node Count:

In [5]:
# 4.2.2
db.count_nodes('Airport')

Airports
0      3214

Node Attributes:

In [6]:
db.node_attributes()

nodeType         nodeLabels                propertyName  \
0   :`Airport`:`Europe`  [Airport, Europe]                          Id   
1   :`Airport`:`Europe`  [Airport, Europe]                    latitude   
2   :`Airport`:`Europe`  [Airport, Europe]                     country   
3   :`Airport`:`Europe`  [Airport, Europe]                   longitude   
4   :`Airport`:`Europe`  [Airport, Europe]                       Label   
5   :`Airport`:`Europe`  [Airport, Europe]                   continent   
6   :`Airport`:`Europe`  [Airport, Europe]                      Degree   
7   :`Airport`:`Europe`  [Airport, Europe]                   Page_Rank   
8   :`Airport`:`Europe`  [Airport, Europe]                 Betweenness   
9   :`Airport`:`Europe`  [Airport, Europe]        Closeness_Centrality   
10  :`Airport`:`Europe`  [Airport, Europe]                   In_Degree   
11  :`Airport`:`Europe`  [Airport, Europe]                  Out_Degree   
12  :`Airport`:`Europe`  [Airport, Europe]              Degree_Airport   
13  :`Airport`:`Europe`  [Airport, Europe]           In_Degree_Airport   
14  :`Airport`:`Europe`  [Airport, Europe]          Out_Degree_Airport   
15  :`Airport`:`Europe`  [Airport, Europe]           Page_Rank_Airport   
16  :`Airport`:`Europe`  [Airport, Europe]         Betweenness_Airport   
17           :`Airport`          [Airport]                          Id   
18           :`Airport`          [Airport]                    latitude   
19           :`Airport`          [Airport]                     country   
20           :`Airport`          [Airport]                   longitude   
21           :`Airport`          [Airport]                       Label   
22           :`Airport`          [Airport]                   continent   
23           :`Airport`          [Airport]                      Degree   
24           :`Airport`          [Airport]                   Page_Rank   
25           :`Airport`          [Airport]                 Betweenness   
26           :`Airport`          [Airport]        Closeness_Centrality   
27           :`Airport`          [Airport]                   In_Degree   
28           :`Airport`          [Airport]                  Out_Degree   
29           :`Airport`          [Airport]              Degree_Airport   
30           :`Airport`          [Airport]           In_Degree_Airport   
31           :`Airport`          [Airport]          Out_Degree_Airport   
32           :`Airport`          [Airport]           Page_Rank_Airport   
33           :`Airport`          [Airport]         Betweenness_Airport   
34    :`Airport`:`Asia`    [Airport, Asia]                          Id   
35    :`Airport`:`Asia`    [Airport, Asia]                    latitude   
36    :`Airport`:`Asia`    [Airport, Asia]                     country   
37    :`Airport`:`Asia`    [Airport, Asia]                   longitude   
38    :`Airport`:`Asia`    [Airport, Asia]                       Label   
39    :`Airport`:`Asia`    [Airport, Asia]                   continent   
40    :`Airport`:`Asia`    [Airport, Asia]                      Degree   
41    :`Airport`:`Asia`    [Airport, Asia]                   Page_Rank   
42    :`Airport`:`Asia`    [Airport, Asia]                 Betweenness   
43    :`Airport`:`Asia`    [Airport, Asia]        Closeness_Centrality   
44    :`Airport`:`Asia`    [Airport, Asia]                   In_Degree   
45    :`Airport`:`Asia`    [Airport, Asia]                  Out_Degree   
46    :`Airport`:`Asia`    [Airport, Asia]              Page_Rank_Asia   
47    :`Airport`:`Asia`    [Airport, Asia]   Closeness_Centrality_Asia   
48    :`Airport`:`Asia`    [Airport, Asia]            Betweenness_Asia   
49    :`Airport`:`Asia`    [Airport, Asia]                     Louvain   
50    :`Airport`:`Asia`    [Airport, Asia]           Label_Propagation   
51    :`Airport`:`Asia`    [Airport, Asia]                   Triangles   
52    :`Airport`:`Asia`    [Airport, Asia]  localClusteringCoefficient   
53    :`Airport`:`A

Relationship Count:

In [7]:
db.count_relationships('TO')

TO_Count
0     36907

Relationship Attributes:

In [8]:
db.rel_attributes()

relType propertyName  propertyTypes  mandatory
0                      :`TO`     distance       [Double]       True
1    :`PATH_Bogota_Directed`    totalCost       [Double]       True
2    :`PATH_Bogota_Directed`      nodeIds    [LongArray]       True
3    :`PATH_Bogota_Directed`        costs  [DoubleArray]       True
4  :`PATH_Bogota_Undirected`    totalCost       [Double]       True
5  :`PATH_Bogota_Undirected`      nodeIds    [LongArray]       True
6  :`PATH_Bogota_Undirected`        costs  [DoubleArray]       True

#### <font color= #5aafc1> **4.3 Visualización en Gephi** </font>

In [9]:
# Le voy a agregar atributos de latitud y longitud en integer para poder meterlo a Gephi 

query4_3 = f"""
    MATCH (n)
    SET n.latInt = toInteger(n.latitude)
    SET n.longInt = toInteger(n.longitude)
"""

db.custom_query(query4_3)

Empty DataFrame
Columns: []
Index: []

#### <font color= #5aafc1> **4.5 Elige el nodo de la pregunta 4.4 como origen y uno de los posibles aeropuertos alcanzables de la misma pregunta como destino, muestra la ruta completa del origen al destino en un mapa.** </font>

In [10]:
query4_5 = f""" 
    MATCH(a:Airport {{Label:'Villahermosa'}})-[r1]->(b)-[r2]->(c)-[r3]->(m:Airport {{Label:'Delhi'}})
    WHERE r3.distance IS NOT NULL
    RETURN a.Label,r1.distance AS Dist1,b.Label,r2.distance AS Dist2,c.Label,r3.distance AS Dist3,m.Label
    ORDER BY Dist1+Dist2+Dist3 DESC
    LIMIT 10;
"""

db.custom_query(query4_5)

a.Label        Dist1      b.Label         Dist2    c.Label  \
0  Villahermosa  1358.755582      Houston  10681.305092      Tokyo   
1  Villahermosa  1358.755582      Houston  11358.558584      Seoul   
2  Villahermosa  1358.755582      Houston  12947.031367       Doha   
3  Villahermosa  1358.755582      Houston  11569.256010    Beijing   
4  Villahermosa  1358.755582      Houston  13139.523542      Dubai   
5  Villahermosa   678.579085  Mexico City   9209.229377      Paris   
6  Villahermosa   678.579085  Mexico City   9855.892994     Munich   
7  Villahermosa   678.579085  Mexico City   9562.676617  Frankfurt   
8  Villahermosa   678.579085  Mexico City   8913.836258     London   
9  Villahermosa   678.579085  Mexico City   9216.602463  Amsterdam   

         Dist3 m.Label  
0  5915.459645   Delhi  
1  4657.687448   Delhi  
2  2552.069255   Delhi  
3  3813.207294   Delhi  
4  2186.045810   Delhi  
5  6572.730383   Delhi  
6  5906.062233   Delhi  
7  6127.480196   Delhi  
8  6738.831316   Delhi  
9  6370.585848   Delhi

#### <font color= #5aafc1> **4.6 Elige un país, que no sea México, analiza el # de aeropuertos que tiene, cuál de ellos es el más importante tomando al menos dos medidas de centralidad para avalar tu respuesta.** </font>

#### <font color= #5aafc1> **4.7 Muestra la mejor ruta entre el aeropuerto de la pregunta 4.6 y el aeropuerto de la CDMX** </font>

#### <font color= #5aafc1> **4.8. Imagina que vives en el país que elegiste en el punto 4.7 (Colombia) y tienes la posibilidad de viajar a cualquier parte del mundo desde su aeropuerto más importante.** </font>

Primero encontraremos el aeropuerto más importante de Colombia en base a la medida de centralidad _Page Rank_, que evalúa las conexiones que salen del nodo y les asigna una relevancia.

In [11]:
query1 = """MATCH (a:Airport)
WHERE a.country = 'Colombia'
RETURN a.Label AS Airport, a.country AS Country, a.Page_Rank AS Page_Rank
ORDER BY Page_Rank DESC
LIMIT 1
"""

db.custom_query(query1)

Airport   Country  Page_Rank
0  Bogota  Colombia   9.420864

El aeropuerto más importante de Colombia es el de Bogotá, que se ubica en la capital.

#### <font color= #5aafc1> &nbsp; &nbsp; **4.8.1. ¿Cuántos kilómetros se requieren para visitar las 7 maravillas del mundo?** </font>
#### <font color= #5aafc1> &nbsp; &nbsp; **4.8.2. ¿Cuál sería la mejor ruta para visitarlas todas, en qué orden y por qué?** </font>

Decidimos trabajar con las siguientes 7 maravillas del mundo modernas:

- Giza Pyramids: Cairo
- Great Wall of China: Beijing
- Colosseum: Rome
- Chichen Itzá: Mérida
- Machu Picchu: Cuzco
- Taj Mahal: Lucknow
- Christ the Redeemer: Rio de Janeiro

Para dar respuesta a las preguntas, se analizaron todas las combinaciones posibles desde Bogotá hacia los destinos mencionados. Entonces, a partir de Bogotá, la mejor ruta para recorrer las 7 maravillas del mundo moderno es:

In [12]:
db.seven_wonders_best_route('Airport','Label','Bogota','distance')

Order  \
0  (Bogota, Merida, Cuzco, Rio De Janeiro, Rome, Cairo, Lucknow, Beijing)   

       Distance  \
0  29588.070071   

                                                                                                                                                                                                                                                                                                               Full_route  
0  [Bogota, Caracas, Merida, Merida, Caracas, Lima, Cuzco, Cuzco, La Paz, Santa Cruz, Sao Paulo, Rio De Janeiro, Rio De Janeiro, Belo Horizonte, Fortaleza, Praia, Santiago Island, Casablanca, Rome, Rome, Cairo, Cairo, Amman, Delhi, Lucknow, Lucknow, Delhi, Kathmandu, Lhasa, Yinchuan, Dongsheng, Beijing, Beijing]

Esto requiere alrededor de 29,588 km.

#### <font color= #5aafc1> &nbsp; &nbsp; **4.8.3. Aeropuertos visitables desde el Aeropuerto más importante de Colombia en menos de 20,000km** </font>

Crearemos un atributo que nos permita evaluar todas las rutas posibles desde Bogotá:

In [13]:
db.delta_pathing('Airport', 'Label', 'Bogota', 'distance')

Directed Delta Pathing for Bogota already exists. Creation skipped.


Source  \
0  Bogota   
1  Bogota   
2  Bogota   

                                                                Bridges  \
0       [Goroka, Port Moresby, Panama City, Nandi, Bogota, Los Angeles]   
1       [Madang, Port Moresby, Panama City, Nandi, Bogota, Los Angeles]   
2  [Mount Hagen, Port Moresby, Panama City, Nandi, Bogota, Los Angeles]   

                                                                                             Partial_Costs  \
0   [0.0, 755.8972047746067, 5606.6853979173375, 14508.500729366064, 17901.63346808019, 18326.70207030612]   
1  [0.0, 755.8972047746067, 5606.6853979173375, 14508.500729366064, 17901.63346808019, 18398.968325213587]   
2  [0.0, 755.8972047746067, 5606.6853979173375, 14508.500729366064, 17901.63346808019, 18417.507717458266]   

     Total_Cost       Target  
0  18326.702070       Goroka  
1  18398.968325       Madang  
2  18417.507717  Mount Hagen

Undirected Delta Pathing for Bogota already exists. Creation skipped.


Source  \
0  Bogota   
1  Bogota   
2  Bogota   

                                                                Bridges  \
0       [Goroka, Port Moresby, Panama City, Nandi, Bogota, Los Angeles]   
1       [Madang, Port Moresby, Panama City, Nandi, Bogota, Los Angeles]   
2  [Mount Hagen, Port Moresby, Panama City, Nandi, Bogota, Los Angeles]   

                                                                                             Partial_Costs  \
0   [0.0, 755.8972047746067, 5606.6853979173375, 14508.500729366064, 17901.63346808019, 18326.70207030612]   
1  [0.0, 755.8972047746067, 5606.6853979173375, 14508.500729366064, 17901.63346808019, 18398.968325213587]   
2  [0.0, 755.8972047746067, 5606.6853979173375, 14508.500729366064, 17901.63346808019, 18417.507717458266]   

     Total_Cost       Target  
0  18326.702070       Goroka  
1  18398.968325       Madang  
2  18417.507717  Mount Hagen

Los aeropuertos y rutas visitables que sumen menos de 20,000 km:

In [14]:
query2 = """MATCH (source:Airport) -[r:PATH_Bogota_Directed]-> (target:Airport)
WHERE r.totalCost <= 20_000 
RETURN COUNT(DISTINCT target.Label) AS Distinct_Visitable_Airports
"""
 
db.custom_query(query2)

Distinct_Visitable_Airports
0                         3001

In [15]:
query3 = """MATCH (source:Airport) -[r:PATH_Bogota_Directed]-> (target:Airport)
MATCH (v:Airport)
WHERE r.totalCost <= 20_000 AND id(v) IN r.nodeIds
RETURN source.Label AS Source, target.Label AS Target, COLLECT(v.Label) AS Bridges, 
target.country AS Target_Country, r.totalCost AS Kilometers
LIMIT 20
"""
 
db.custom_query(query3)

Source              Target  \
0   Bogota              Goroka   
1   Bogota              Madang   
2   Bogota         Mount Hagen   
3   Bogota              Nadzab   
4   Bogota        Port Moresby   
5   Bogota               Wewak   
6   Bogota        Narssarssuaq   
7   Bogota            Godthaab   
8   Bogota         Sondrestrom   
9   Bogota               Thule   
10  Bogota            Akureyri   
11  Bogota         Egilsstadir   
12  Bogota          Isafjordur   
13  Bogota            Keflavik   
14  Bogota           Reykjavik   
15  Bogota  Sault Sainte Marie   
16  Bogota         St. Anthony   
17  Bogota         Baie Comeau   
18  Bogota          Bagotville   
19  Bogota          Baker Lake   

                                                                                      Bridges  \
0                             [Goroka, Port Moresby, Panama City, Nandi, Bogota, Los Angeles]   
1                             [Madang, Port Moresby, Panama City, Nandi, Bogota, Los Angeles]   
2                        [Mount Hagen, Port Moresby, Panama City, Nandi, Bogota, Los Angeles]   
3                             [Nadzab, Port Moresby, Panama City, Nandi, Bogota, Los Angeles]   
4                                     [Port Moresby, Panama City, Nandi, Bogota, Los Angeles]   
5                              [Port Moresby, Wewak, Panama City, Nandi, Bogota, Los Angeles]   
6                           [Narssarssuaq, Godthaab, Keflavik, Santo Domingo, Bogota, Boston]   
7                                         [Godthaab, Keflavik, Santo Domingo, Bogota, Boston]   
8                            [Godthaab, Sondrestrom, Keflavik, Santo Domingo, Bogota, Boston]   
9   [Godthaab, Thule, Keflavik, Santo Domingo, Bogota, Boston, Ilulissat, Qaanaaq, Upernavik]   
10                   [Godthaab, Akureyri, Keflavik, Reykjavik, Santo Domingo, Bogota, Boston]   
11                [Godthaab, Egilsstadir, Keflavik, Reykjavik, Santo Domingo, Bogota, Boston]   
12                 [Godthaab, Isafjordur, Keflavik, Reykjavik, Santo Domingo, Bogota, Boston]   
13                                                  [Keflavik, Santo Domingo, Bogota, Boston]   
14                             [Godthaab, Keflavik, Reykjavik, Santo Domingo, Bogota, Boston]   
15                                                      [Sault Sainte Marie, Toronto, Bogota]   
16                                      [St. Anthony, Halifax, Goose Bay, Punta Cana, Bogota]   
17                                                      [Baie Comeau, Quebec, Bogota, Newark]   
18                                                   [Bagotville, Montreal, Bogota, New York]   
19                         [Baker Lake, Rankin Inlet, Winnipeg, Bogota, Atlanta, Minneapolis]   

      Target_Country    Kilometers  
0   Papua New Guinea  18326.702070  
1   Papua New Guinea  18398.968325  
2   Papua New Guinea  18417.507717  
3   Papua New Guinea  18226.120224  
4   Papua New Guinea  17901.633468  
5   Papua New Guinea  18662.934623  
6          Greenland  10016.244656  
7          Greenland   9552.032638  
8          Greenland   9869.219679  
9          Greenland  11300.657141  
10           Iceland  11231.414101  
11           Iceland  11362.315578  
12           Iceland  11203.541950  
13           Iceland   8149.907903  
14           Iceland  10981.729726  
15            Canada   4866.658445  
16            Canada   5960.885142  
17            Canada   5073.708911  
18            Canada   4919.210857  
19            Canada   7216.236215

Algunos de los destinos visitables desde Bogotá a menos de 20,000 km de distancia son Papua New Guinea, Greenland, Iceland y Canada. Las rutas para llegar a cada destino están especificadas en la tabla de arriba, mostrando los países a recorrer.

## <font color= #5aafc1> **Tours** </font>

#### <font color= #93c47d> &nbsp; &nbsp; **Tour 1: The Grand Tour (Europa)** </font>

In [16]:
# Subgrafo específico de aeropuertos en Europa
db.create_subgraph_continent('Airport','TO','Europe')

Successfully created Airport_Europe Subgraph with Airport nodes and TO relationships!


In [17]:
# 1
db.custom_query(f"""
    CALL gds.degree.write('Airport_Europe', {{writeProperty: 'Out_Degree_Europe'}} )
    YIELD centralityDistribution, nodePropertiesWritten
    RETURN centralityDistribution.min AS minimumScore, 
    centralityDistribution.mean AS meanScore, 
    nodePropertiesWritten;
""")

Error with query: {code: Neo.ClientError.Procedure.ProcedureCallFailed} {message: Failed to invoke procedure `gds.degree.write`: Caused by: org.neo4j.gds.core.loading.GraphNotFoundException: Graph with name `Airport_Europe` does not exist on database `neo5j`. It might exist on another database.}


In [18]:
db.custom_query("""MATCH (a:Airport {continent:'Europe'})
                WHERE a.Out_Degree_Europe IS NOT null
                RETURN a.Label AS Name, a.Out_Degree_Europe AS ODE
                ORDER BY ODE DESC
                LIMIT 10;""")

Empty DataFrame
Columns: [Name, ODE]
Index: []

In [19]:
# 2
db.custom_query(f"""
    CALL gds.pageRank.write(
        'Airport_Europe',
        {{writeProperty: 'Page_Rank_Europe'}} )
        YIELD nodePropertiesWritten, ranIterations;
""")

Error with query: {code: Neo.ClientError.Procedure.ProcedureCallFailed} {message: Failed to invoke procedure `gds.pageRank.write`: Caused by: org.neo4j.gds.core.loading.GraphNotFoundException: Graph with name `Airport_Europe` does not exist on database `neo5j`. It might exist on another database.}


In [20]:
top_europe_countries = ['Germany','United Kingdom','France','Italy','Spain','Ukraine','Poland','Romania','Netherlands','Belgium']

for i in top_europe_countries:
    result = db.custom_query(f"""
    MATCH (a:Airport {{country: '{i}'}})
    WHERE a.Page_Rank_Europe IS NOT NULL
    RETURN a.Label AS Airport, a.Page_Rank_Europe AS PR
    ORDER BY PR DESC
    LIMIT 1;
    """)

    print(f"{i}: {result['Airport']}")


Germany: Series([], Name: Airport, dtype: object)
United Kingdom: Series([], Name: Airport, dtype: object)
France: Series([], Name: Airport, dtype: object)
Italy: Series([], Name: Airport, dtype: object)
Spain: Series([], Name: Airport, dtype: object)
Ukraine: Series([], Name: Airport, dtype: object)
Poland: Series([], Name: Airport, dtype: object)
Romania: Series([], Name: Airport, dtype: object)
Netherlands: Series([], Name: Airport, dtype: object)
Belgium: Series([], Name: Airport, dtype: object)


In [40]:
# 3
cities = ['Munich','London','Paris','Rome','Barcelona','Kiev','Warsaw','Bucharest','Amsterdam','Brussels']

for i in cities:
    result = db.temp_dijkstra_directed('Airport','Label','Guadalajara',i,'distance')
    print(f"""
    Ruta GDL-{i}: {result["Distance"]} \n
    Middle Airports: {result['Bridges']}
""")


    Ruta GDL-Munich: 0    10031.953936
Name: Distance, dtype: float64 

    Middle Airports: 0    [Guadalajara, Houston, Munich]
Name: Bridges, dtype: object


    Ruta GDL-London: 0    3358.698645
1    9086.463951
2    9126.420797
3    9131.317161
4    9094.852864
5    9123.639416
Name: Distance, dtype: float64 

    Middle Airports: 0                     [Guadalajara, Chicago, London]
1    [Guadalajara, Houston, Toronto, Dublin, London]
2           [Guadalajara, Houston, Montreal, London]
3    [Guadalajara, Houston, Toronto, Dublin, London]
4                     [Guadalajara, Houston, London]
5    [Guadalajara, Houston, Toronto, Dublin, London]
Name: Bridges, dtype: object


    Ruta GDL-Paris: 0    9403.563969
1    9407.824968
Name: Distance, dtype: float64 

    Middle Airports: 0              [Guadalajara, Houston, Paris]
1    [Guadalajara, Houston, Montreal, Paris]
Name: Bridges, dtype: object


    Ruta GDL-Rome: 0    10517.777807
1    10460.905875
Name: Distance, dtype: float6

Estas rutas son complejas y difíciles de tratar para una persona turista, por ende se propone la ruta de GDL a Londres via la CDMX, la cuál es apenas 300 km más larga pero es logísticamente más sencilla

In [ ]:
db.custom_query(f"MATCH (a:Airport {{Label:'Guadalajara'}})-[r:TO]->(a2:Airport {{Label: 'Mexico City'}})-[r2:TO]->(a3:Airport {{Label: 'London'}}) RETURN r.distance + r2.distance, a3.Id")

r.distance + r2.distance  a3.Id
0               9373.457896    507

De forma similar, para el retorno, se propone la siguiente ruta: Barcelona-GDL via Londres y CDMX

In [ ]:
db.custom_query(f"MATCH (a:Airport {{Label:'Barcelona'}})-[r:TO]->(a2:Airport {{Label: 'London'}})RETURN r.distance, a2.Id") # se va a tomar el aeropuerto con id=507

r.distance  a2.Id
0  1192.097273    492
1  1110.296837    502
2  1148.999108    507
3  1186.942188    548

In [ ]:
# 4
grand_1 = db.custom_query(f"""
    MATCH (a:Airport {{Id:507}})-[r:TO]->(a2:Airport {{Label:'Munich'}}) 
    RETURN r.distance AS Dist;
""") # En este caso, identifico a Londres con su Id específico ya que hay varios aeropuertos con el Label:"Londres". Este es el mismo aeropuerto en el que llegamos

grand_2 = db.custom_query(f"""
    MATCH (a:Airport {{Label:'Munich'}})-[r:TO]->(a2:Airport {{Label:'Paris'}}) 
    RETURN r.distance AS Dist;
""")

grand_3 = db.custom_query(f"""
    MATCH (a:Airport {{Label:'Paris'}})-[r:TO]->(a2:Airport {{Label:'Rome'}}) 
    RETURN r.distance AS Dist;
""")

grand_4 = db.custom_query(f"""
    MATCH (a:Airport {{Label:'Rome'}})-[r:TO]->(a2:Airport {{Label:'Barcelona'}}) 
    RETURN r.distance AS Dist;
""")

grand_f = db.temp_dijkstra_directed('Airport','Label','Barcelona','Guadalajara','distance')

print(f"""
    Grand Tour Report: \n
    Total Distance: {9373+grand_1['Dist']+grand_2['Dist']+grand_3['Dist']+grand_4['Dist']+(1186+9373)}
    Cost Per Leg: 
""")


    Grand Tour Report: 

    Total Distance: 0    23536.266188
1             NaN
Name: Dist, dtype: float64
    Cost Per Leg: 



In [ ]:
db.temp_dijkstra_directed('Airport','Label','Barcelona','Guadalajara','distance')

Source       Target                  Middle     Distance
0  Barcelona  Guadalajara  [570, 1854, 1809, 818]  9746.264626
1  Barcelona  Guadalajara  [1329, 1340, 832, 818]  4296.911075

In [ ]:
# 5
db.custom_query(f"""CALL gds.louvain.write(
            'Airport_Europe',
            {{writeProperty: 'Louvain_Europe'}} )
            YIELD communityCount, modularity, modularities;
            """)

communityCount  modularity  \
0               8    0.267822   

                                                     modularities  
0  [0.2541399296152414, 0.26615043547199513, 0.26782181796958543]

In [ ]:
db.custom_query(f"""
    MATCH (a:Airport {{continent:'Europe'}})
    RETURN a.Louvain_Europe AS LV, COLLECT(DISTINCT a.country) AS Countries
""")

LV  \
0    1.0   
1   45.0   
2   66.0   
3  135.0   
4  128.0   
5  457.0   
6  314.0   
7    NaN   
8  477.0   

                                                                                                                                                                                                                            Countries  
0                                                                                                                                                                                                                           [Iceland]  
1                                       [Iceland, Estonia, Finland, United Kingdom, Netherlands, Denmark, Faroe Islands, Norway, Poland, Sweden, Bulgaria, Cyprus, Azerbaijan, Ukraine, Belarus, Latvia, Lithuania, Armenia, Georgia]  
2                                                                                           [Belgium, Germany, United Kingdom, Netherlands, Spain, Albania, France, Greece, Italy, Czech Republic, Portugal, Switzerland, Montenegro]  
3  [Belgium, Germany, Finland, United Kingdom, Netherlands, Norway, Poland, Sweden, Bulgaria, Croatia, Spain, France, Greece, Hungary, Italy, Czech Republic, Bosnia and Herzegovina, Romania, Moldova, Macedonia, Slovakia, Ukraine]  
4                                                                                     [Germany, United Kingdom, Guernsey, Jersey, Isle of Man, Ireland, Poland, Sweden, Spain, Cyprus, France, Malta, Portugal, Gibraltar, Lithuania]  
5                                                             [Germany, Luxembourg, Spain, Bulgaria, Croatia, Greece, Italy, Slovenia, Austria, Bosnia and Herzegovina, Romania, Switzerland, Serbia, Montenegro, Slovakia, Portugal]  
6                                                                                                                                                                                                                    [United Kingdom]  
7                                                                                                                                                      [Italy, Czech Republic, Romania, Armenia, Cyprus, France, Ukraine, Azerbaijan]  
8                                                                                                                                                                                                                          [Portugal]

#### <font color= #d04e4e> &nbsp; &nbsp; **Tour 2: The Rising Sun (Asia)** </font>

Crear una nueva label para Asia:

In [22]:
query1 = """MATCH (a:Airport)
WHERE a.continent = 'Asia'
SET a:Asia
RETURN a.Label, a.country, a.continent
LIMIT 10;
"""
db.custom_query(query1)

a.Label a.country a.continent
0   Tel-aviv    Israel        Asia
1       Elat    Israel        Asia
2      Haifa    Israel        Asia
3       Ovda    Israel        Asia
4   Tel-aviv    Israel        Asia
5     Ankara    Turkey        Asia
6      Adana    Turkey        Asia
7    Antalya    Turkey        Asia
8  Gaziantep    Turkey        Asia
9      Konya    Turkey        Asia

Subgrafos para el continente Asiático:

In [14]:
db.create_subgraph('Asia', 'TO', 'NATURAL', 'distance')
db.create_subgraph('Asia', 'TO', 'UNDIRECTED', 'distance')

Successfully created Directed Subgraph with Asia nodes and TO relationships!
Successfully created Undirected Subgraph with Asia nodes and TO relationships!


In [24]:
db.check_subgraphs()

graphName  nodeCount  relationshipCount database  \
0  Undirected_Airport       3214              73814    neo5j   
1       Directed_Asia        885               9492    neo5j   
2    Reversed_Airport       3214              36907    neo5j   
3      Airport_Europe        503               9777    neo5j   
4    Directed_Airport       3214              36907    neo5j   
5     Undirected_Asia        885              18984    neo5j   

                          creationTime  
0  2025-07-16T16:34:42.281446900-06:00  
1  2025-07-16T16:38:00.553513400-06:00  
2  2025-07-16T16:34:41.984945900-06:00  
3  2025-07-16T16:35:03.980411300-06:00  
4  2025-07-16T16:34:41.583769300-06:00  
5  2025-07-16T16:38:00.665171300-06:00

#### <font color= #d04e4e> &nbsp; &nbsp; **Page Rank en Asia** </font>

In [25]:
db.page_rank('Asia')

Page Rank Attribute added Successfully!


In [26]:
query2 = """MATCH (a:Asia)
RETURN a.Label AS Airport, a.country AS Country, a.Page_Rank_Asia AS Page_Rank_Asia
ORDER BY Page_Rank_Asia DESC
LIMIT 10
"""
db.custom_query(query2)

Airport               Country  Page_Rank_Asia
0        Moscow                Russia       11.220566
1       Beijing                 China       10.682839
2      Istanbul                Turkey        9.571835
3     Guangzhou                 China        8.093310
4      Shanghai                 China        7.933281
5         Dubai  United Arab Emirates        7.749905
6     Singapore             Singapore        7.725984
7  Kuala Lumpur              Malaysia        7.700118
8        Manila           Philippines        7.654879
9         Tokyo                 Japan        7.357576

#### <font color= #d04e4e> &nbsp; &nbsp; **Betwenness Centrality en Asia** </font>

In [27]:
db.betwenness('Asia')

Betweenness Attribute added Successfully!


In [ ]:
query3 = """MATCH (a:Asia)
RETURN a.Label AS Airport, a.country AS Country, a.Betweenness_Asia AS Betweenness_Asia
ORDER BY Betweenness_Asia DESC
LIMIT 10
"""
db.custom_query(query3)

Airport               Country  Betweenness_Asia
0       Beijing                 China      91986.036674
1      Istanbul                Turkey      74337.337881
2        Moscow                Russia      65191.001742
3         Dubai  United Arab Emirates      64763.590727
4     Singapore             Singapore      62140.411718
5  Kuala Lumpur              Malaysia      52975.485438
6        Manila           Philippines      50537.595223
7       Jakarta             Indonesia      48881.531825
8       Bangkok              Thailand      47495.081555
9     Guangzhou                 China      44691.190370

#### <font color= #d04e4e> &nbsp; &nbsp; **Recomendaciones en Asia** </font>

In [7]:
countries = ['Moscow', 'Beijing', 'Istanbul', 'Guangzhou', 'Shanghai']
for i in countries:
    query = f"""MATCH (a:Asia) -[r:TO]-> (a2:Asia)
    WHERE a.Label = '{i}' AND a2.country <> a.country
    RETURN a.Label AS Start_Point, a2.country AS Country,
    COUNT(DISTINCT a2.Label) AS Airport_Count
    ORDER BY Airport_Count DESC
    LIMIT 3;
    """
    display(db.custom_query(query))

Start_Point     Country  Airport_Count
0      Moscow  Uzbekistan             11
1      Moscow  Kazakhstan             11
2      Moscow       China              5

Start_Point      Country  Airport_Count
0     Beijing       Russia             10
1     Beijing  South Korea              6
2     Beijing        Japan              5

Start_Point       Country  Airport_Count
0    Istanbul        Russia             14
1    Istanbul  Saudi Arabia              7
2    Istanbul          Iran              6

Start_Point   Country  Airport_Count
0   Guangzhou     Japan              3
1   Guangzhou    Taiwan              3
2   Guangzhou  Thailand              3

Start_Point      Country  Airport_Count
0    Shanghai        Japan             17
1    Shanghai  South Korea              6
2    Shanghai       Taiwan              3

#### <font color= #d04e4e> &nbsp; &nbsp; **Triangle Count en Asia** </font>

In [28]:
db.community_algorithms('Asia')

Louvain Attribute added Successfully!
Label Propagation Attribute added Successfully!
Triangle Count Attribute added Successfully!
Local Clustering Coefficient Attribute added Successfully!
Strongly CC Attribute added Successfully!
Weakly CC Attribute added Successfully!
K1-Coloring Attribute added Successfully!


In [29]:
db.color_correction('Asia')

Louvain Attribute Color Corrected Successfully!
Label_Propagation Attribute Color Corrected Successfully!
Triangles Attribute Color Corrected Successfully!
localClusteringCoefficient Attribute Color Corrected Successfully!
Community_SCC Attribute Color Corrected Successfully!
Community_WCC Attribute Color Corrected Successfully!
K1Color Attribute Color Corrected Successfully!


In [34]:
query = """MATCH (a:Asia)
RETURN a.Triangles_Asia AS Triangle_Communities, 
COUNT(DISTINCT a.Label) as Count, COLLECT(DISTINCT a.Label) as Airports
ORDER BY Count DESC
LIMIT 3;
"""
db.custom_query(query)

Triangle_Communities  Count  \
0                     1    223   
1                     2    138   
2                     4     57   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          Airports  
0  [Ovda, Konya, Merzifon, Denizli, Rafha, Turaif, Wejh, Sanandaj, Zabol, Fujeirah, Khasab, Faisalabad, Gilgit, Panjgur, Zhob, Fengnin, Tainan, Nanki-shirahama, Obihiro, Iki, Tsushima, Monbetsu, Fukue, Kitakyushu, Izumo, Tottori, Tokushima, Hachijojima, Oshima, Kumejima, Cotabato, San Jose, Pevek, Barnaul, Kemorovo, Bhuj, Belgaum, Bhaunagar, Jamnagar, Rajkot, Trinciomalee, Cox's Bazar, Jessore, Kulu, Gwalior, Kanpur, Ludhiaha, Bhairawa, Pokhara, Biratnagar, Agatti Island, Vijayawada, Lampang, Phrae, Narathiwat, Ranong, Trang, Sakon Nakhon, Loei, Kyaukpyu, Myeik, Putao, Wamena, Merauke, Poso, Luwuk, Babo, Limbang, Lahad Datu, Ketapang, Alor Setar, Kerteh, Ipoh, Hotan, Wajima, Naha, Tanjung Pinang, Ende, Labuhan Bajo, Waingapu, Nakhon Si Thammarat, Nakhon Phanom, Ubon Ratchathani, Khon Kaen, Sukhothai, Ataq, Al Ghaidah Intl, Socotra, Nevsehir, Tagbilaran, Phon Savan, Mae Hong Son, Sihanoukville, Roi Et, Buri Ram, Trat, Luang Namtha, Muang Xay, Rach Gia, Ca Mau, ...]  
1           [Elat, Haifa, Tel-aviv, Guriat, Hafr Al-batin, Arar, Sharurah, Bandar Mahshahr, Bandar Lengeh, Birjand, Ramsar, Aqaba, Abu Dhabi, Moenjodaro, Matsumoto, Nakashibetsu, Wakkanai, Yamaguchi, Amami, Kochi, Yamagata, Misawa, Kwangju, Yeosu, Ulsan, Pohang, Pagadian, Tacloban, Dumaguete, Caticlan, Stavropol, Magnetiogorsk, Aurangabad, Baroda, Jabalpur, Khajuraho, Porbandar, Surat, Udaipur, Rajshahi, Saidpur, Allahabad, Dehra Dun, Jodhpur, Madurai, Pattaya, Phitsanulok, Hue, Kengtung, Myitkyina, Sittwe, Thandwe, Biak, Nabire, Gorontalo, Pontianak, Jambi, Bengkulu, Kendari, Malang, Tanjung Pandan, Palangkaraya, Maumere, Bima, Redang, Nan, Dong Hoi, Chu Lai, Tuy Hoa, Busuanga, Butuan, Legazpi, Ozamis, Kaadedhdhoo, Diu, Agri, Adiyaman, Dawadmi, Al-Jawf, Noshahr, Ardabil, Chitral, Dalbandin, Fukushima, Sacheon, Sangley Point, Kokshetau, Petropavlosk, Zhezkazgan, Anadyr, Nyagan, Kirov, Kurgan, Termez, Cheboksary, Saransk, Hubli, Dimapur, Laamu Atoll, Haiphong, ...]  
2                                                                                                                                                                                                                                                                                                                                                                                                                                          [Herat, Al-ahsa, El-baha, Bisha, Nejran, Abadan, Gwadar, Rahim Yar Khan, Kinmen, Makung, Memanbetsu, Asahikawa, Hanamaki, Miyako, Romblon, Zamboanga, Uralsk, Pavlodar, Semiplatinsk, Abakan, Bhopal, Agartala, Aizwal, Silchar, Chandigarh, Leh, Bagan, Ternate, Labuan, Banda Aceh, Kuantan, Pyongyang, Xichang, Kashi, Hodeidah, Dalat, Dipolog, Arkhangelsk, Saratov, Uromiyeh, Dera Ghazi Khan, Kushiro, Kostanay, Izhevsk, Andizhan, Khanabad, Pleiku, Bandar Lampung-Sumatra Island, Quzhou, Qiqihar, Dan

#### <font color= #d04e4e> &nbsp; &nbsp; **Dijkstra en Asia** </font>

In [34]:
db.find_best_route('Airport', 'Label', 'Guadalajara', ['Beijing', 'Chongqing', 'Tokyo', 'Singapore', 'Bangkok', 'Seoul', 'Yangon'], 'distance')

Order  \
0  (Guadalajara, Tokyo, Seoul, Beijing, Chongqing, Yangon, Bangkok, Singapore)   

       Distance  \
0  18617.800872   

                                                                                                                                                                         Full_route  
0  [Guadalajara, Los Angeles, Tokyo, Tokyo, Seoul, Seoul, Dalian, Chifeng, Beijing, Beijing, Chongqing, Chongqing, Kunming, Yangon, Yangon, Bangkok, Bangkok, Singapore, Singapore]

In [35]:
db.temp_dijkstra_directed('Airport', 'Label', 'Singapore', 'Guadalajara', 'distance')

Source       Target                                       Bridges  \
0  Singapore  Guadalajara  [Singapore, Tokyo, Los Angeles, Guadalajara]   

       Distance  
0  16235.260875

#### <font color= #d04e4e> &nbsp; &nbsp; **Ruta más económica** </font>

In [36]:
db.find_best_route('Airport', 'Label', 'Guadalajara', ['Beijing', 'Tokyo', 'Seoul'], 'distance')

Order      Distance  \
0  (Guadalajara, Tokyo, Seoul, Beijing)  13373.295402   

                                                                                  Full_route  
0  [Guadalajara, Los Angeles, Tokyo, Tokyo, Seoul, Seoul, Dalian, Chifeng, Beijing, Beijing]

In [37]:
db.temp_dijkstra_directed('Airport', 'Label', 'Beijing', 'Guadalajara', 'distance')

Source       Target  \
0  Beijing  Guadalajara   
1  Beijing  Guadalajara   

                                                         Bridges      Distance  
0                       [Beijing, Seattle, Phoenix, Guadalajara]  12147.603112  
1  [Beijing, Chifeng, Shenyang, Vancouver, Phoenix, Guadalajara]  12399.258093